File: frontend/src/utils/reasoningBullets.ts
const MIN_BULLETS = 1;
const MAX_BULLETS = 5;

const SENTENCE_PATTERN = /[^.!?]+[.!?]+/g;

function splitSentences(text: string): string[] {
  const sentences: string[] = [];
  let match: RegExpExecArray | null;
  const regex = new RegExp(SENTENCE_PATTERN.source, 'g');
  while ((match = regex.exec(text)) !== null) {
    const sentence = match[0].trim();
    if (sentence) sentences.push(sentence);
  }
  return sentences;
}

/**
 * Split outfit reasoning into 3–5 bullet-friendly lines for display.
 */
export function reasoningToBullets(reasoning: string): string[] {
  const trimmed = reasoning.trim();
  if (!trimmed) return [];

  const byNewline = trimmed
    .split(/\n+/)
    .map((s) => s.replace(/^[-•*]\s*/, '').trim())
    .filter(Boolean);

  if (trimmed.includes('\n') && byNewline.length >= MIN_BULLETS) {
    return byNewline.slice(0, MAX_BULLETS);
  }

  const bySentence = splitSentences(trimmed);
  if (bySentence.length >= MIN_BULLETS) {
    return bySentence.slice(0, MAX_BULLETS);
  }

  return [trimmed];
}

File: ios-client/OutfitSuggestor/Utils/ReasoningBullets.swift
//
//  ReasoningBullets.swift
//  OutfitSuggestor
//

import Foundation

enum ReasoningBullets {
    private static let minBullets = 1
    private static let maxBullets = 5

    /// Split outfit reasoning into bullet-friendly lines (3–5 when possible).
    static func toBullets(_ reasoning: String) -> [String] {
        let trimmed = reasoning.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        let byNewline = trimmed
            .components(separatedBy: .newlines)
            .map { $0.replacingOccurrences(of: #"^[-•*]\s*"#, with: "", options: .regularExpression) }
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        if trimmed.contains("\n"), byNewline.count >= minBullets {
            return Array(byNewline.prefix(maxBullets))
        }

        let pattern = #"[^.!?]+[.!?]+"#
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let range = NSRange(trimmed.startIndex..., in: trimmed)
            let matches = regex.matches(in: trimmed, range: range)
            let sentences = matches.compactMap { match -> String? in
                guard let r = Range(match.range, in: trimmed) else { return nil }
                return String(trimmed[r]).trimmingCharacters(in: .whitespacesAndNewlines)
            }.filter { !$0.isEmpty }

            if sentences.count >= minBullets {
                return Array(sentences.prefix(maxBullets))
            }
        }

        return [trimmed]
    }
}
